
## Final Project Submission

Please fill out:
* Student name: Kevin Gitahi
* Student pace: full time student
* Scheduled project review date/time: 
* Instructor name: Nikita Njoroge
* Blog post URL:


## PHASE_3 PROJECT

## TITLE:MACHINE LEARNING CLASSIFICATION MODEL FOR PREDICTING CUSTOMER CHURN AT SYRIA TEL

## 1.0 BUSINESS UNDERSTANDING/PROJECT OVERVIEW
#### 1.1 INTRODUCTION.

Client churn, also known as customer churn or subscriber churn, refers to the loss of customers or subscribers over a specific period of time.

Fighting client churn has become a crucial concern in the dynamic telecommunications sector, as customers have many options for service providers and frequently switch carriers.Currently,the industry's typical yearly churn rate in the US ranges from 11% to 25% according to latest report (year 2020) by statista https://www.statista.com/statistics/816735/customer-churn-rate-by-industry-us/, and keeping current clients has become crucial. Acknowledging the significant financial difference between bringing in new clients and keeping existing clients, seasoned operators make keeping existing clients their top priority.

Understanding and reducing client churn is essential for businesses because:
* Churn directly translates to lost recurring revenue. Retaining existing clients is generally cheaper than acquiring new ones.
* High churn rates can indicate customer dissatisfaction and potentially harm the company's brand image.
* By analyzing churn data, companies can identify patterns and risk factors associated with churn. This allows them to develop strategies to retain at-risk customers.

Combating Client Churn:
Companies can implement various strategies to reduce churn:
* Customer Satisfaction Surveys: Understanding customer pain points and satisfaction levels helps to identify areas for improvement.
* Loyalty Programs: Reward programs and incentives can encourage customers to stay.
* Personalized Offers: Tailoring offers and promotions to individual customer needs can increase value perception.
* Improved Customer Service: Providing excellent customer support builds trust and loyalty.
* Proactive Engagement: Identifying at-risk customers and offering targeted interventions can prevent churn.

By focusing on customer satisfaction, value delivery, and proactive engagement, Telecom companies can effectively reduce client churn and maintain a healthy customer base.


#### 1.2 PROBLEM STATEMENT.

SyriaTel, a telecommunications company, has experienced a significant rise in customer churn rates within the US market during the past financial period. This churn is leading to customer loss and negatively impacting SyriaTel's revenue. To address this challenge,I aim to develop a machine learning model that will help SyriaTel company's stakeholders to identify customers at risk of churn in the US.Additionally,this model will be used to develop targeted customer retention strategies and mitigate churn.

#### 1.3 MAIN OBJECTIVE.

To develop a machine learning classification model that accurately predicts customer churn in the US market for SyriaTel, enabling them to identify customers at high risk of churning and implement targeted customer retention strategies.
 
#### 1.4 SPECIFIC OBJECTIVES.
* To identify the key factors that contribute to customer churn in Syria Tel.
* To develop a model that will accurately predict the probability of churn for individual customers in the US market.
* To determine the likely solution to lower churn.
 



## 2.DATA UNDERSTANDING

The dataset contains various features related to telecom customer behavior, service usage, and account information. 
The dataset has details such as:
1. The customer's state
2.  Account length
3.  Area code
4.  Phone number
5.  International plan
6.  Voice mail plan
7.  Number of voicemail messages
8.  Total duration and charges for calls made during the day, evening, and night.
9.  It also includes information on international calls
10. Customer service calls
11. Whether or not the customer churned (terminated their contract).

The data will be suitable to build a classifier to predict whether a customer will ("soon") stop doing business with SyriaTel telecommunications company.

The data is obtained from kaggle website.

Data Source : https://www.kaggle.com/datasets/becksddf/churn-in-telecoms-dataset 


The data set is made up of 21 columns made up of different features as shown below:

Summary of Features in the Dataset
* state: the state the customer lives in - different states in America
* account length: the number of days since the customer possessed an account
* area code: the area code of the customer
* phone number: the phone number of the customer
* international plan: true if the customer has the international plan, otherwise false
* voice mail plan: true if the customer has the voice mail plan, otherwise false
* number vmail messages: the number of voicemails the customer has sent
* total day minutes: total number of minutes the customer has been on calls during the day
* total day calls: total number of calls the user has made during the day
* total day charge: total amount of money the customer was charged by the Telecom company for calls made during the day
* total eve minutes: total number of minutes the customer has been on calls in the evening
* total eve calls: total number of calls the customer has made in the evening
* total eve charge: total amount of money the customer was charged by the Telecom company for calls made in the evening
* total night minutes: total number of minutes the customer has been on calls during the night
* total night calls: total number of calls the customer has done during the night
* total night charge: total amount of money the customer was charged by the Telecom company for calls made during the night
* total intl minutes: total number of minutes the user has been on international calls
* total intl calls: total number of international calls the customer has made
* total intl charge: total amount of money the customer was charged by the Telecom company for international calls
* customer service calls: number of calls the customer has made to customer service
* churn: target variable which is true if the customer terminated their contract, otherwise false

## 3.DATA PREPARATION(DATA CLEANING)
#### 3.1 Data preparation
#### Import the necessary libraries

In [91]:
#importing libraries for data handling
import pandas as pd
import numpy as np

#importing libraries for visualization
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

#importing libraries for modeling
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.linear_model import LogisticRegression
from sklearn.datasets import make_classification
from sklearn.metrics import confusion_matrix
from sklearn.tree import DecisionTreeClassifier 
from sklearn.metrics import accuracy_score, roc_curve, auc
from sklearn.preprocessing import OneHotEncoder
from sklearn.tree import export_graphviz, plot_tree
from IPython.display import Image
from sklearn import tree
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import GridSearchCV
import warnings
warnings.filterwarnings("ignore")

In [92]:
# loading data
data = pd.read_csv("data.csv")
data

state  account length  area code phone number international plan  \
0       KS             128        415     382-4657                 no   
1       OH             107        415     371-7191                 no   
2       NJ             137        415     358-1921                 no   
3       OH              84        408     375-9999                yes   
4       OK              75        415     330-6626                yes   
...    ...             ...        ...          ...                ...   
3328    AZ             192        415     414-4276                 no   
3329    WV              68        415     370-3271                 no   
3330    RI              28        510     328-8230                 no   
3331    CT             184        510     364-6381                yes   
3332    TN              74        415     400-4344                 no   

     voice mail plan  number vmail messages  total day minutes  \
0                yes                     25              265.1   
1                yes                     26              161.6   
2                 no                      0              243.4   
3                 no                      0              299.4   
4                 no                      0              166.7   
...              ...                    ...                ...   
3328             yes                     36              156.2   
3329              no                      0              231.1   
3330              no                      0              180.8   
3331              no                      0              213.8   
3332             yes                     25              234.4   

      total day calls  total day charge  ...  total eve calls  \
0                 110             45.07  ...               99   
1                 123             27.47  ...              103   
2                 114             41.38  ...              110   
3                  71             50.90  ...               88   
4                 113             28.34  ...              122   
...               ...               ...  ...              ...   
3328               77             26.55  ...              126   
3329               57             39.29  ...               55   
3330              109             30.74  ...               58   
3331              105             36.35  ...               84   
3332              113             39.85  ...               82   

      total eve charge  total night minutes  total night calls  \
0                16.78                244.7                 91   
1                16.62                254.4                103   
2                10.30                162.6                104   
3                 5.26                196.9                 89   
4                12.61                186.9                121   
...                ...                  ...                ...   
3328             18.32                279.1                 83   
3329             13.04                191.3                123   
3330             24.55                191.9                 91   
3331             13.57                139.2                137   
3332             22.60                241.4                 77   

      total night charge  total intl minutes  total intl calls  \
0                  11.01                10.0                 3   
1                  11.45                13.7                 3   
2                   7.32                12.2                 5   
3                   8.86                 6.6                 7   
4                   8.41                10.1                 3   
...                  ...                 ...               ...   
3328               12.56                 9.9                 6   
3329                8.61                 9.6                 4   
3330                8.64                14.1                 6   
3331                6.26                 5.0                10   
3332               10.86           

In [93]:
#create a function that:
#loads data
#looks at the columns
#check for the information,df.info()
#check for data types
#check for the first five rows

def explore_data(file):
    data = pd.read_csv(file)
    print("Data loaded successfully!")
  # Explore data
    print("\nColumn names:")
    print(data.columns)
    print("\nData information:")
    print(data.info())
    print("\nData types:")
    print(data.dtypes)
    print("\nFirst 5 rows:")
    return data.head()

In [94]:
file = "data.csv"

In [95]:
explore_data(file)

Data loaded successfully!

Column names:
Index(['state', 'account length', 'area code', 'phone number',
       'international plan', 'voice mail plan', 'number vmail messages',
       'total day minutes', 'total day calls', 'total day charge',
       'total eve minutes', 'total eve calls', 'total eve charge',
       'total night minutes', 'total night calls', 'total night charge',
       'total intl minutes', 'total intl calls', 'total intl charge',
       'customer service calls', 'churn'],
      dtype='object')

Data information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   3333 non-null   object 
 1   account length          3333 non-null   int64  
 2   area code               3333 non-null   int64  
 3   phone number            3333 non-null   object 
 4   international plan      3333 non

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  

[5 rows x 21 columns]

#### Observations
From the information above obtained from the dataset ,I was able to observe the following:

In the data cleaning stage,

I will remove the white spaces in column names,to minimize the spaces and make the names readable

* The total number of entries was 3333
* I had both Numerical and categorical data:
  
          a)Numerical data( int and float data types)
  
* account length                                  
* area code
* total day minutes                               
* number vmail messages
* total day charge         
* total eve minutes        
* total eve calls             
* total eve charge          
* total night minutes     
* total night calls        
* total night charge       
* total intl minutes       
* total intl calls           
* total intl charge         
* customer service calls                         

             
          b)categorical data(object data type ,boolean)
  
* state
* phone number             
* international plan         
* voice mail plan           
* churn

Churn is my target variable and is of data type boolean "True" to show if the customer churn and "False" to show customer did not churn

I did not have any missing value since all the numbers of entries matched (3333)


In [96]:
#create a function that will:
#check for last rows
#check for the shape of dataset
#check for the summary of the dataset .describe() function

def get_last_rows():
    return data.tail()

def get_data_shape():
    return data.shape

def get_summary_statistics():
    return data.describe()

  # Return the results
    return get_last_rows(), get_data_shape(), get_summary_statistics()

In [97]:
get_last_rows()

state  account length  area code phone number international plan  \
3328    AZ             192        415     414-4276                 no   
3329    WV              68        415     370-3271                 no   
3330    RI              28        510     328-8230                 no   
3331    CT             184        510     364-6381                yes   
3332    TN              74        415     400-4344                 no   

     voice mail plan  number vmail messages  total day minutes  \
3328             yes                     36              156.2   
3329              no                      0              231.1   
3330              no                      0              180.8   
3331              no                      0              213.8   
3332             yes                     25              234.4   

      total day calls  total day charge  ...  total eve calls  \
3328               77             26.55  ...              126   
3329               57             39.29  ...               55   
3330              109             30.74  ...               58   
3331              105             36.35  ...               84   
3332              113             39.85  ...               82   

      total eve charge  total night minutes  total night calls  \
3328             18.32                279.1                 83   
3329             13.04                191.3                123   
3330             24.55                191.9                 91   
3331             13.57                139.2                137   
3332             22.60                241.4                 77   

      total night charge  total intl minutes  total intl calls  \
3328               12.56                 9.9                 6   
3329                8.61                 9.6                 4   
3330                8.64                14.1                 6   
3331                6.26                 5.0                10   
3332               10.86                13.7                 4   

      total intl charge  customer service calls  churn  
3328               2.67                       2  False  
3329               2.59                       3  False  
3330               3.81                       2  False  
3331               1.35                       2  False  
3332               3.70                       0  False  

[5 rows x 21 columns]

In [98]:
get_data_shape()

(3333, 21)

In [99]:
get_summary_statistics()

account length    area code  number vmail messages  total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       total day calls  total day charge  total eve minutes  total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       total eve charge  total night minutes  total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       total night charge  total intl minutes  total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       total intl charge  customer service calls  
count        3333.000000             3333.000000  
mean            2.764581                1.562856  
std             0.753773                1.315491  
min             0.000000                0.000000  
25%             2.300000                1.000000  
50%             2.780000                1.000000  
75%             3.270000                2.000000  
max             5.400000                9.000000

#### Observations
I observed that the data is made up of 21 columns and 3333 rows

I also had an output of the data summmary statistics that shows the count,mean,standard deviation,minimum and maximum values and also percentiles for the different columns of the dataset that contained numerical data.

#### 3.2 Data cleaning
When carrying out data cleaning i had to ask myself the following Questions so that I could  clean my data appropriately.
    
    QUESTIONS
 * Are there missing values?
 * Any white spaces?
 * Any duplicates?

In [100]:
# Update column names to include underscore make them readable
updated_columns = []
for column_names in data.columns:
    updated_columns.append(column_names.replace(' ', '_'))
data.columns = updated_columns
data.columns

Index(['state', 'account_length', 'area_code', 'phone_number',
       'international_plan', 'voice_mail_plan', 'number_vmail_messages',
       'total_day_minutes', 'total_day_calls', 'total_day_charge',
       'total_eve_minutes', 'total_eve_calls', 'total_eve_charge',
       'total_night_minutes', 'total_night_calls', 'total_night_charge',
       'total_intl_minutes', 'total_intl_calls', 'total_intl_charge',
       'customer_service_calls', 'churn'],
      dtype='object')

checking for missing values to accertin that we don"t have any NAN values

In [101]:
#check for missing values
data.isnull().sum()

state                     0
account_length            0
area_code                 0
phone_number              0
international_plan        0
voice_mail_plan           0
number_vmail_messages     0
total_day_minutes         0
total_day_calls           0
total_day_charge          0
total_eve_minutes         0
total_eve_calls           0
total_eve_charge          0
total_night_minutes       0
total_night_calls         0
total_night_charge        0
total_intl_minutes        0
total_intl_calls          0
total_intl_charge         0
customer_service_calls    0
churn                     0
dtype: int64

From the output we can confidently say that we do not have missing values in our dataset.

Note that both state and area_code are representation of a geographical area,we can go ahead and check the state column and calculates the frequency of each unique value within the 'state' column and area code.

In [102]:
data.state.value_counts()

state
WV    106
MN     84
NY     83
AL     80
WI     78
OH     78
OR     78
WY     77
VA     77
CT     74
MI     73
ID     73
VT     73
TX     72
UT     72
IN     71
MD     70
KS     70
NC     68
NJ     68
MT     68
CO     66
NV     66
WA     66
RI     65
MA     65
MS     65
AZ     64
FL     63
MO     63
NM     62
ME     62
ND     62
NE     61
OK     61
DE     61
SC     60
SD     60
KY     59
IL     58
NH     56
AR     55
GA     54
DC     54
HI     53
TN     53
AK     52
LA     51
PA     45
IA     44
CA     34
Name: count, dtype: int64

In [103]:
data.area_code.value_counts()

area_code
415    1655
510     840
408     838
Name: count, dtype: int64

we can go ahead and drop the are_code column .it does not give us a lot of information.

In [104]:
data = data.drop("area_code" , axis =1 )

In [105]:
data.columns

Index(['state', 'account_length', 'phone_number', 'international_plan',
       'voice_mail_plan', 'number_vmail_messages', 'total_day_minutes',
       'total_day_calls', 'total_day_charge', 'total_eve_minutes',
       'total_eve_calls', 'total_eve_charge', 'total_night_minutes',
       'total_night_calls', 'total_night_charge', 'total_intl_minutes',
       'total_intl_calls', 'total_intl_charge', 'customer_service_calls',
       'churn'],
      dtype='object')

next we check for duplicates

In [106]:
data.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
3328    False
3329    False
3330    False
3331    False
3332    False
Length: 3333, dtype: bool

The data had no duplicates

We go ahead and check for outliers

In [ ]:
# checking for outliers in numerical columns
numeric_cols = data.select_dtypes(include=['int64', 'float64'])
num_plots = len(numeric_cols.columns)
num_rows = (num_plots + 3) // 4 # 4 columns
num_cols = min(num_plots, 4)  

plt.figure(figsize=(12, 3 * num_rows))

for i, col in enumerate(numeric_cols.columns):
    plt.subplot(num_rows, num_cols, i + 1)
    data[col].plot(kind='box')
    plt.title(f'Boxplot of {col}')  # Set title
    plt.ylabel(col)  # Set y-label

plt.tight_layout()
plt.show()

Our data looks  all good and we can go to the next steps of analysis and modeling.

## 4.DATA ANALYSIS

## 5.MODELING

## 6.EVALUATION

## 7.CONCLUSION & RECOMMENDATIONS

## 8.NEXT STEPS